### PR TITLE
Adding functionality to expose Dual Values

### PR DESCRIPTION
**Pull request by dandxy89**
As per your suggestion on Reddit I had a stab at implementing the functionality to expose dual values.

Appears to be failing build due to Upstream...

### CODE REVIEW

**Comment by dandxy89** [comment]
Also includes some minor formatting - VSCode seems eagar!

File: src/dual.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+//! the constraint is equal to its bound. Also known as the shadow price
+//!
+
+/// A problem dual values (shadow prices)
+pub trait DualValues {
+    /// Get all dual values for a given problem
+    fn get_dual_values(&self) -> &[f64];
+    /// Get the dual value of a problem by index
+    fn get_dual_value(&self, constraint_index: usize) -> f64;
+}
```
Comment (lovasoa) [inline src/dual.rs:12]: I don't think the trait should work like that. This forces either the solver's internal representation to always have the dual values as a continuous slice of doubles, or us to always compute the dual values when we don't need them. 

I think the trait should allow code to look like the following : 

```rust
let my_constraint : ConstraintReference = pb.add_constraint( x + y << 7)
let solution = pb.solve()?
let constraint_dual =  solution.get_dual().value(my_constraint)
```

This means : 
  - creating `add_constraint(&mut self, Constraint) -> ConstraintReference`
  - creating a `SolutionWithDual` which allows computing the dual values, and a `DualValues` trait (or just a struct) that takes a  `ConstraintReference` and returns the corresponding dual value.

File: tests/shadow_price.rs [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#[allow(clippy::float_cmp)]
```
Comment (lovasoa) [inline tests/shadow_price.rs:1]: Not a good idea 😬 . We already have float_eq in the devdependencies, you should use that: https://docs.rs/float_eq/0.5.0/float_eq/macro.assert_float_eq.html

File: tests/solver_scaling.rs [hunk 3]
```diff
@@ -3,6 +3,7 @@ use good_lp::{constraint, default_solver, variable, variables, Expression, Solut
 
 const BIG_NUM: usize = 1000; // <- Set this higher to test how good_lp and the solvers scale
 
+#[allow(clippy::needless_range_loop)]
```
Comment (lovasoa) [inline tests/solver_scaling.rs:6]: Clippy was right, we should change the loop to `v.iter().enumerate()`

**Review by lovasoa — CHANGES_REQUESTED** [review]
Thank you very much ! I added my remarks below.
The failing tests highlight a serious memory error, we need to investigate that.

**Comment by lovasoa** [comment]
I reported the memory corruption to https://github.com/KardinalAI/coin_cbc/issues/9

File: tests/shadow_price.rs [hunk 4]
```diff
@@ -63,12 +70,15 @@ mod tests {
...
 
-        assert_eq!(solution.eval(&objective) as u32, 4100);
-        assert_eq!(solution.value(n_chairs) as u32, 29);
-        assert_eq!(solution.value(n_tables) as u32, 40);
-        assert_eq!(
-            solution.get_dual_values(),
-            vec![-15.000000000000004, -4.999999999999992]
+        assert_float_eq!(4100.0, solution.eval(&objective), abs <= 1e-10);
+        assert_float_eq!(29.999999999999996, solution.value(n_chairs), abs <= 1e-10);
+        assert_float_eq!(40.00000000000001, solution.value(n_tables), abs <= 1e-10);
```
Comment (lovasoa) [inline tests/shadow_price.rs:75]: This should probably be 30.0 and 40.0 instead. You can reduce the abs if necessary

File: tests/shadow_price.rs [hunk 5]
```diff
@@ -31,13 +32,19 @@ mod tests {
...
-        assert_eq!(solution.eval(&objective), 77.30220492866408);
-        assert_eq!(solution.value(x1), 20.363164721141374);
-        assert_eq!(solution.value(x2), 6.485084306095981);
-        assert_eq!(solution.get_dual_values()[3], solution.get_dual_value(3));
-        assert_eq!(
-            solution.get_dual_values(),
-            vec![-0.0, -0.0, -0.6809338521400778, -0.09208819714656294]
+        assert_float_eq!(77.30220492866408, solution.eval(&objective), abs <= 1e-10);
+        assert_float_eq!(20.363164721141374, solution.value(x1), abs <= 1e-10);
+        assert_float_eq!(6.485084306095981, solution.value(x2), abs <= 1e-10);
```
Comment (lovasoa) [inline tests/shadow_price.rs:37]: Can we choose a problem with more "round" solutions ? These are easier to check.

File: tests/shadow_price.rs [hunk 6]
```diff
@@ -63,12 +70,15 @@ mod tests {
...
-        assert_eq!(solution.value(n_tables) as u32, 40);
-        assert_eq!(
-            solution.get_dual_values(),
-            vec![-15.000000000000004, -4.999999999999992]
+        assert_float_eq!(4100.0, solution.eval(&objective), abs <= 1e-10);
+        assert_float_eq!(29.999999999999996, solution.value(n_chairs), abs <= 1e-10);
+        assert_float_eq!(40.00000000000001, solution.value(n_tables), abs <= 1e-10);
+
+        assert_float_eq!(
+            -15.000000000000004,
```
Comment (lovasoa) [inline tests/shadow_price.rs:78]: 15

File: tests/shadow_price.rs [hunk 7]
```diff
@@ -63,12 +70,15 @@ mod tests {
...
+            solution.get_dual_value(0),
+            abs <= 1e-10
         );
+        assert_float_eq!(-4.999999999999992, solution.get_dual_value(1), abs <= 1e-10);
```
Comment (lovasoa) [inline tests/shadow_price.rs:82]: 5

**Comment by dandxy89** [comment]
@lovasoa had a stab at implementing your initial suggestions. It works but I suspect you've got a better handle of Rust than me...

Comments about rounding and probelm are valid - will update once you're happy with the actual solution for dual variables.

File: src/solvers/minilp.rs [hunk 8]
```diff
@@ -70,6 +78,27 @@ impl SolverModel for MiniLpProblem {
...
+            false => minilp::ComparisonOp::Le,
+        };
+        let constant = -c.expression.constant;
+        self.problem.add_constraint(&coefficients, op, constant);
+        self.n_constraints += 1;
+
+        ConstraintReference {
+            index: self.n_constraints - 1,
+        }
+    }
```
Comment (lovasoa) [inline src/solvers/minilp.rs:101]: We probably shouldn't duplicate code between `add_constraint` and `with`. We can have a default method implementation in `SolverModel` for with that internally uses add_constraint.

**Review by lovasoa — CHANGES_REQUESTED** [review]

File: src/solvers/mod.rs [hunk 9]
```diff
@@ -123,3 +126,12 @@ impl<N: Into<f64> + Clone> Solution for HashMap<Variable, N> {
...
     }
 }
+
+/// The dual value measures the increase in the objective function's value per unit
+/// increase in the variable's value. The dual value for a constraint is nonzero only when
+/// the constraint is equal to its bound. Also known as the shadow price
+///
+pub trait SolutionWithDual: Solution {
+    /// Get a dual value of a problem for a given ConstraintReference
+    fn get_dual_value(&self, c: ConstraintReference) -> f64;
```
Comment (lovasoa) [inline src/solvers/mod.rs:136]: If the underlying solver only offers a function to compute all dual variables together, this forces to allocate an array for all the dual results, and then throw away everything but the variable you are interested in.

What I was suggesting was something that would let you write

```
let constraint_dual : f64 =  solution.get_dual().value(my_constraint)
```

That is, first compute all the dual values, then extract the one you are interested in.

```rust
```

**Review by lovasoa — CHANGES_REQUESTED** [review]

**Comment by dandxy89** [comment]
@lovasoa had another go at attempting to implement what you've advised. 

I feel it could be better but I think I'm reaching my _current knowledge boundary_ of Rust...Happy for you to jump in a re-work. 

Keen to learn!

**Comment by dandxy89** [comment]
@lovasoa do you have an opportunity to review and provide guidance?

**Comment by lovasoa** [comment]
Hello ! Sorry, I have been a little busy, but I'm going to look into this today !

**Comment by lovasoa** [comment]
@dandxy89 : I made the changes I wanted. If it looks good to you, I'll merge and publish a new version.

**Review by dandxy89 — COMMENTED** [review]
Awesome. Thanks @lovasoa

**Comment by lovasoa** [comment]
Published as 0.4.

Would you like to be added to the rust-or org ?

**Comment by dandxy89** [comment]
@lovasoa sure.

**Pull request merged**